File: platforms/android/app/src/main/java/io/foxbiz/shellular/browser/EmbeddedProxyServer.java
package io.foxbiz.shellular.browser;

import android.os.Handler;
import android.os.Looper;
import android.util.Base64;
import android.util.Log;
import android.webkit.WebView;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class EmbeddedProxyServer {

    private static final String TAG = "EmbeddedProxy";
    public static final EmbeddedProxyServer INSTANCE = new EmbeddedProxyServer();

    public static class ParsedRequest {
        public final String method;
        public final String url;
        public final Map<String, String> headers;
        public final byte[] body;

        public ParsedRequest(String method, String url, Map<String, String> headers, byte[] body) {
            this.method = method;
            this.url = url;
            this.headers = headers;
            this.body = body;
        }
    }

    public static class PendingConnection {
        public final Socket socket;
        public final OutputStream outputStream;
        public final String requestId;
        public boolean useChunked = false;

        public PendingConnection(Socket socket, OutputStream outputStream, String requestId) {
            this.socket = socket;
            this.outputStream = outputStream;
            this.requestId = requestId;
        }
    }

    private final ExecutorService threadPool = Executors.newCachedThreadPool();
    private final Map<Integer, ServerSocket> listeners = new ConcurrentHashMap<>();
    private final Set<Integer> activePorts = ConcurrentHashMap.newKeySet();
    private final Map<String, PendingConnection> pendingConnections = new ConcurrentHashMap<>();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    private WebView mainWebView;

    public void setMainWebView(WebView webView) {
        this.mainWebView = webView;
    }

    public boolean startServer(int port) {
        if (activePorts.contains(port)) return true;

        try {
            ServerSocket serverSocket = new ServerSocket();
            serverSocket.setReuseAddress(true);
            serverSocket.bind(new InetSocketAddress("127.0.0.1", port));
            listeners.put(port, serverSocket);
            activePorts.add(port);

            threadPool.execute(() -> acceptLoop(serverSocket, port));
            Log.d(TAG, "Server started on 127.0.0.1:" + port);
            return true;
        } catch (IOException e) {
            Log.e(TAG, "Failed to start server on port " + port + ": " + e.getMessage());
            return false;
        }
    }

    public void stopServer(int port) {
        ServerSocket ss = listeners.remove(port);
        activePorts.remove(port);
        if (ss != null) {
            try { ss.close(); } catch (IOException ignored) {}
        }
    }

    public void stopAll() {
        for (Map.Entry<Integer, ServerSocket> entry : listeners.entrySet()) {
            try { entry.getValue().close(); } catch (IOException ignored) {}
            activePorts.remove(entry.getKey());
        }
        listeners.clear();
    }

    public boolean hasServer(int port) {
        return activePorts.contains(port);
    }

    public void cancelAllPending() {
        for (PendingConnection pc : pendingConnections.values()) {
            try { pc.socket.close(); } catch (IOException ignored) {}
        }
        pendingConnections.clear();
    }

    private void acceptLoop(ServerSocket serverSocket, int port) {
        while (!serverSocket.isClosed()) {
            try {
                Socket socket = serverSocket.accept();
                threadPool.execute(() -> handleConnection(socket));
            } catch (IOException e) {
                if (!serverSocket.isClosed()) {
                    Log.w(TAG, "Accept error on port " + port + ": " + e.getMessage());
                }
                break;
            }
        }
    }

    private void handleConnection(Socket socket) {
        try {
            socket.setSoTimeout(30000);
            InputStream in = socket.getInputStream();
            OutputStream out = socket.getOutputStream();

            ParsedRequest request = readHTTPRequest(in);
            if (request == null) {
                sendErrorResponse(out, 400, "Bad Request");
                socket.close();
                return;
            }

            String requestId = "req_" + UUID.randomUUID().toString().replace("-", "");
            pendingConnections.put(requestId, new PendingConnection(socket, out, requestId));

            proxyRequest(requestId, request);
        } catch (IOException e) {
            try { socket.close(); } catch (IOException ignored) {}
        }
    }

    private ParsedRequest readHTTPRequest(InputStream in) throws IOException {
        ByteArrayOutputStream headerBuffer = new ByteArrayOutputStream();
        int prev3 = -1, prev2 = -1, prev1 = -1;
        int b;
        while ((b = in.read()) != -1) {
            headerBuffer.write(b);
            if (prev3 == '\r' && prev2 == '\n' && prev1 == '\r' && b == '\n') {
                break;
            }
            prev3 = prev2;
            prev2 = prev1;
            prev1 = b;
        }

        byte[] headerBytes = headerBuffer.toByteArray();
        if (headerBytes.length == 0) return null;
        int headerEnd = headerBytes.length - 4;

        String headerStr = new String(headerBytes, 0, headerEnd, StandardCharsets.UTF_8);
        String[] lines = headerStr.split("\r\n");

        if (lines.length == 0) return null;
        String[] requestParts = lines[0].split(" ");
        if (requestParts.length < 2) return null;

        String method = requestParts[0].toUpperCase();
        String url = requestParts[1];

        Map<String, String> headers = new HashMap<>();
        for (int i = 1; i < lines.length; i++) {
            int colonIdx = lines[i].indexOf(':');
            if (colonIdx > 0) {
                String key = lines[i].substring(0, colonIdx).trim();
                String value = lines[i].substring(colonIdx + 1).trim();
                String existing = headers.get(key);
                if (existing != null && key.equalsIgnoreCase("Set-Cookie")) {
                    headers.put(key, existing + "\n" + value);
                } else {
                    headers.put(key, value);
                }
            }
        }

        byte[] body = null;
        String contentLengthStr = headers.get("Content-Length");
        if (contentLengthStr != null) {
            try {
                int contentLength = Integer.parseInt(contentLengthStr.trim());
                if (contentLength > 0) {
                    ByteArrayOutputStream bodyBuffer = new ByteArrayOutputStream();
                    byte[] chunk = new byte[Math.min(contentLength, 8192)];
                    int totalRead = 0;
                    while (totalRead < contentLength) {
                        int n = in.read(chunk, 0, Math.min(contentLength - totalRead, chunk.length));
                        if (n == -1) break;
                        bodyBuffer.write(chunk, 0, n);
                        totalRead += n;
                    }
                    body = bodyBuffer.toByteArray();
                }
            } catch (NumberFormatException ignored) {}
        }

        return new ParsedRequest(method, url, headers, body);
    }

    private void proxyRequest(String requestId, ParsedRequest request) {
        WebView wv = mainWebView;
        if (wv == null) {
            sendResponseError(requestId, "WebView not available");
            return;
        }

        mainHandler.post(() -> {
            String headersJson = mapToJson(request.headers);
            String bodyStr = request.body != null ? new String(request.body, StandardCharsets.UTF_8) : "";

            String escapedHeaders = escapeForJS(headersJson);
            String escapedBody = escapeForJS(bodyStr);
            String escapedUrl = escapeForJS(request.url);
            String escapedMethod = escapeForJS(request.method);

            String js = "window.__shellularProxy&&window.__shellularProxy.httpRequest('"
                    + requestId + "','"
                    + escapedMethod + "','"
                    + escapedUrl + "','"
                    + escapedHeaders + "','"
                    + escapedBody + "')";

            wv.evaluateJavascript(js, null);
        });
    }

    public void sendResponseStart(String requestId, int status, String statusText, String headersJson) {
        PendingConnection pc = pendingConnections.get(requestId);
        if (pc == null) return;

        try {
            Writer writer = new OutputStreamWriter(pc.outputStream, StandardCharsets.UTF_8);
            writer.write("HTTP/1.1 " + status + " " + statusText + "\r\n");

            Map<String, String> headers = jsonToMap(headersJson);
            boolean hasTransferEncoding = false;
            boolean hasContentType = false;

            for (Map.Entry<String, String> entry : headers.entrySet()) {
                String lower = entry.getKey().toLowerCase();
                if (lower.equals("connection")) continue;

                if (lower.equals("content-type")) {
                    hasContentType = true;
                    Log.d(TAG, "Response Content-Type: " + entry.getValue() + " for " + requestId);
                }

                if (lower.equals("transfer-encoding")) {
                    hasTransferEncoding = true;
                    if (entry.getValue().toLowerCase().contains("chunked")) {
                        pc.useChunked = true;
                    }
                }

                if (lower.equals("set-cookie")) {
                    String[] cookies = entry.getValue().split("\n");
                    for (String cookie : cookies) {
                        String trimmed = cookie.trim();
                        if (!trimmed.isEmpty()) {
                            writer.write(entry.getKey() + ": " + trimmed + "\r\n");
                        }
                    }
                } else {
                    writer.write(entry.getKey() + ": " + entry.getValue() + "\r\n");
                }
            }

            if (!hasContentType) {
                Log.w(TAG, "Missing Content-Type for " + requestId + " headersJson=" + headersJson.substring(0, Math.min(200, headersJson.length())));
            }

            Log.d(TAG, "Response " + requestId + " status=" + status + " headers=" + headers.size()
                    + " chunked=" + pc.useChunked + " hasContentType=" + hasContentType);

            if (!hasTransferEncoding) {
                writer.write("Connection: close\r\n");
            }

            writer.write("\r\n");
            writer.flush();
        } catch (IOException e) {
            Log.w(TAG, "Error writing response start: " + e.getMessage());
            cleanupConnection(requestId);
        }
    }

    public void sendResponseData(String requestId, String chunkBase64, int index) {
        PendingConnection pc = pendingConnections.get(requestId);
        if (pc == null) return;

        try {
            byte[] chunk = Base64.decode(chunkBase64, Base64.DEFAULT);

            if (chunk != null && chunk.length > 0) {
                Log.d(TAG, "Response data " + requestId + " index=" + index + " size=" + chunk.length);
            } else if (chunk == null) {
                Log.w(TAG, "Response data " + requestId + " index=" + index + " base64 decode returned null");
                return;
            }

            if (pc.useChunked) {
                String hexSize = Integer.toHexString(chunk.length);
                pc.outputStream.write((hexSize + "\r\n").getBytes(StandardCharsets.UTF_8));
                pc.outputStream.write(chunk);
                pc.outputStream.write("\r\n".getBytes(StandardCharsets.UTF_8));
            } else {
                pc.outputStream.write(chunk);
            }
            pc.outputStream.flush();
        } catch (IOException e) {
            Log.w(TAG, "Error writing response data: " + e.getMessage());
            cleanupConnection(requestId);
        }
    }

    public void sendResponseEnd(String requestId) {
        PendingConnection pc = pendingConnections.remove(requestId);
        if (pc == null) return;

        try {
            if (pc.useChunked) {
                pc.outputStream.write("0\r\n\r\n".getBytes(StandardCharsets.UTF_8));
                pc.outputStream.flush();
            }
            pc.outputStream.flush();
            pc.socket.close();
        } catch (IOException ignored) {}
    }

    public void sendResponseError(String requestId, String errorMessage) {
        PendingConnection pc = pendingConnections.remove(requestId);
        if (pc == null) return;

        try {
            byte[] body = errorMessage.getBytes(StandardCharsets.UTF_8);
            Writer writer = new OutputStreamWriter(pc.outputStream, StandardCharsets.UTF_8);
            writer.write("HTTP/1.1 502 Bad Gateway\r\n");
            writer.write("Content-Type: text/plain\r\n");
            writer.write("Content-Length: " + body.length + "\r\n");
            writer.write("Connection: close\r\n");
            writer.write("\r\n");
            writer.flush();
            pc.outputStream.write(body);
            pc.outputStream.flush();
            pc.socket.close();
        } catch (IOException e) {
            cleanupConnection(requestId);
        }
    }

    private void cleanupConnection(String requestId) {
        PendingConnection pc = pendingConnections.remove(requestId);
        if (pc != null) {
            try { pc.socket.close(); } catch (IOException ignored) {}
        }
    }

    private void sendErrorResponse(OutputStream out, int status, String message) throws IOException {
        byte[] body = message.getBytes(StandardCharsets.UTF_8);
        Writer writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
        writer.write("HTTP/1.1 " + status + " Error\r\n");
        writer.write("Content-Type: text/plain\r\n");
        writer.write("Content-Length: " + body.length + "\r\n");
        writer.write("Connection: close\r\n");
        writer.write("\r\n");
        writer.flush();
        out.write(body);
        out.flush();
    }

    private static String escapeForJS(String s) {
        return s.replace("\\", "\\\\")
                .replace("'", "\\'")
                .replace("\n", "\\n")
                .replace("\r", "\\r");
    }

    private static String mapToJson(Map<String, String> map) {
        org.json.JSONObject obj = new org.json.JSONObject();
        for (Map.Entry<String, String> entry : map.entrySet()) {
            try {
                obj.put(entry.getKey(), entry.getValue());
            } catch (Exception ignored) {}
        }
        return obj.toString();
    }

    private static Map<String, String> jsonToMap(String json) {
        Map<String, String> result = new HashMap<>();
        try {
            org.json.JSONObject obj = new org.json.JSONObject(json);
            java.util.Iterator<String> keys = obj.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                result.put(key, obj.getString(key));
            }
        } catch (Exception e) {
            Log.w(TAG, "Error parsing JSON: " + e.getMessage() + " json=" + json.substring(0, Math.min(200, json.length())));
        }
        return result;
    }
}

File: platforms/macos/shellular/EmbeddedProxyServer.swift
import Foundation
import Network
import WebKit

final class EmbeddedProxyServer {
    private static let tunnelFrameBytes = 64 * 1024

    private final class RawTunnel {
        let id: String
        let port: Int
        let connection: NWConnection
        weak var webView: WKWebView?
        var sendCredit = 0
        var reading = false
        var closed = false

        init(id: String, port: Int, connection: NWConnection, webView: WKWebView) {
            self.id = id
            self.port = port
            self.connection = connection
            self.webView = webView
        }
    }

    struct PendingConnection {
        let connection: NWConnection
        let requestId: String
        var useChunked = false
    }

    static let shared = EmbeddedProxyServer()

    private let queue = DispatchQueue(label: "shellular.embeddedProxy", qos: .userInitiated)
    private var listeners: [Int: NWListener] = [:]
    private var tunnelListeners: [String: NWListener] = [:]
    private var tunnelTargets: [Int: String] = [:]
    private var rawTunnels: [String: RawTunnel] = [:]
    private(set) var activePorts: Set<Int> = []
    private var pendingConnections: [String: PendingConnection] = [:]

    func startServer(port: Int, webView: WKWebView) -> Bool {
        guard !activePorts.contains(port) else { return true }

        let params = NWParameters.tcp
        params.allowLocalEndpointReuse = true

        guard let nwPort = NWEndpoint.Port(rawValue: UInt16(port)) else {
            print("[EmbeddedProxy] Invalid port: \(port)")
            return false
        }

        params.requiredLocalEndpoint = NWEndpoint.hostPort(
            host: NWEndpoint.Host("127.0.0.1"),
            port: nwPort
        )

        guard let listener = try? NWListener(using: params) else {
            print("[EmbeddedProxy] Failed to create listener on port \(port)")
            return false
        }

        listener.newConnectionHandler = { [weak self] connection in
            self?.handleConnection(connection, webView: webView)
        }

        listener.stateUpdateHandler = { [weak self] state in
            switch state {
            case .failed(let error):
                print("[EmbeddedProxy] Listener failed on port \(port): \(error)")
                self?.activePorts.remove(port)
            case .cancelled:
                self?.activePorts.remove(port)
            default:
                break
            }
        }

        listener.start(queue: queue)
        listeners[port] = listener
        activePorts.insert(port)
        print("[EmbeddedProxy] Server started on 127.0.0.1:\(port)")
        return true
    }

    func stopServer(port: Int) {
        guard let listener = listeners[port] else { return }
        listener.cancel()
        listeners.removeValue(forKey: port)
        activePorts.remove(port)
    }

    func stopAll() {
        for (_, listener) in listeners {
            listener.cancel()
        }
        listeners.removeAll()
        for (_, listener) in tunnelListeners {
            listener.cancel()
        }
        tunnelListeners.removeAll()
        tunnelTargets.removeAll()
        for (_, tunnel) in rawTunnels {
            tunnel.closed = true
            tunnel.connection.cancel()
        }
        rawTunnels.removeAll()
        activePorts.removeAll()
    }

    func hasServer(port: Int) -> Bool {
        return activePorts.contains(port)
    }

    func startTunnelServer(port: Int, targetHost: String, webView: WKWebView) -> Bool {
        if tunnelTargets[port] == targetHost { return true }
        guard !activePorts.contains(port) else { return false }
        guard let nwPort = NWEndpoint.Port(rawValue: UInt16(port)) else { return false }

        var started = false
        for bindHost in ["127.0.0.1", "::1"] {
            let params = NWParameters.tcp
            params.allowLocalEndpointReuse = true
            params.requiredLocalEndpoint = .hostPort(
                host: NWEndpoint.Host(bindHost),
                port: nwPort
            )
            guard let listener = try? NWListener(using: params) else { continue }
            let key = "\(bindHost):\(port)"
            listener.newConnectionHandler = { [weak self, weak webView] connection in
                guard let self, let webView else {
                    connection.cancel()
                    return
                }
                self.handleRawTunnelConnection(
                    connection,
                    targetHost: targetHost,
                    port: port,
                    webView: webView
                )
            }
            listener.stateUpdateHandler = { [weak self] state in
                guard let self else { return }
                if case .failed = state {
                    self.tunnelListeners.removeValue(forKey: key)
                    if !self.tunnelListeners.keys.contains(where: { $0.hasSuffix(":\(port)") }) {
                        self.tunnelTargets.removeValue(forKey: port)
                        self.activePorts.remove(port)
                    }
                }
            }
            listener.start(queue: queue)
            tunnelListeners[key] = listener
            started = true
        }
        guard started else { return false }
        tunnelTargets[port] = targetHost
        activePorts.insert(port)
        return true
    }

    func stopTunnelServer(port: Int) {
        let suffix = ":\(port)"
        for key in tunnelListeners.keys.filter({ $0.hasSuffix(suffix) }) {
            tunnelListeners.removeValue(forKey: key)?.cancel()
        }
        tunnelTargets.removeValue(forKey: port)
        for tunnel in rawTunnels.values.filter({ $0.port == port }) {
            closeRawTunnel(tunnel, notifyJavaScript: true)
        }
        if listeners[port] == nil {
            activePorts.remove(port)
        }
    }

    func tunnelOpened(id: String, sendCredit: Int) {
        queue.async { [weak self] in
            guard let self, let tunnel = self.rawTunnels[id], !tunnel.closed else { return }
            tunnel.sendCredit = min(1024 * 1024, max(0, sendCredit))
            self.receiveRawTunnelData(tunnel)
        }
    }

    func addTunnelWindow(id: String, bytes: Int) {
        queue.async { [weak self] in
            guard let self, let tunnel = self.rawTunnels[id], !tunnel.closed else { return }
            tunnel.sendCredit = min(1024 * 1024, tunnel.sendCredit + max(0, bytes))
            self.receiveRawTunnelData(tunnel)
        }
    }

    func sendTunnelData(id: String, data: Data) {
        queue.async { [weak self] in
            guard let self, let tunnel = self.rawTunnels[id], !tunnel.closed else { return }
            tunnel.connection.send(content: data, completion: .contentProcessed { [weak self] error in
                guard let self else { return }
                if let error {
                    self.closeRawTunnel(tunnel, notifyJavaScript: true, error: error.localizedDescription)
                    return
                }
                self.callProxy(
                    tunnel.webView,
                    function: "tcpConsumed",
                    arguments: ["id": id, "bytes": data.count]
                )
            })
        }
    }

    func endTunnel(id: String) {
        queue.async { [weak self] in
            guard let self, let tunnel = self.rawTunnels[id], !tunnel.closed else { return }
            tunnel.connection.send(
                content: nil,
                contentContext: .finalMessage,
                isComplete: true,
                completion: .contentProcessed { error in
                    if let error {
                        self.closeRawTunnel(tunnel, notifyJavaScript: true, error: error.localizedDescription)
                    }
                }
            )
        }
    }

    func closeTunnel(id: String, error: String? = nil) {
        queue.async { [weak self] in
            guard let self, let tunnel = self.rawTunnels[id] else { return }
            self.closeRawTunnel(tunnel, notifyJavaScript: false, error: error)
        }
    }

    func cancelAllPending() {
        queue.async { [weak self] in
            guard let self else { return }
            for (_, pending) in self.pendingConnections {
                pending.connection.cancel()
            }
            self.pendingConnections.removeAll()
        }
    }

    func sendResponseStart(requestId: String, status: Int, statusText: String, headersJson: String) {
        queue.async { [weak self] in
            guard let self, var pending = self.pendingConnections[requestId] else { return }

            var headerLines = "HTTP/1.1 \(status) \(statusText)\r\n"
            var hasTransferEncoding = false

            if let data = headersJson.data(using: .utf8),
               let headers = try? JSONSerialization.jsonObject(with: data) as? [String: String] {
                for (key, value) in headers {
                    let lower = key.lowercased()
                    if lower == "connection" { continue }

                    if lower == "transfer-encoding" {
                        hasTransferEncoding = true
                        if value.lowercased().contains("chunked") {
                            pending.useChunked = true
                        }
                    }

                    if lower == "set-cookie" {
                        let cookies = value.components(separatedBy: "\n")
                        for cookie in cookies {
                            let trimmed = cookie.trimmingCharacters(in: .whitespaces)
                            if !trimmed.isEmpty {
                                headerLines += "\(key): \(trimmed)\r\n"
                            }
                        }
                    } else {
                        headerLines += "\(key): \(value)\r\n"
                    }
                }
            }

            if !hasTransferEncoding {
                headerLines += "Connection: close\r\n"
            }

            headerLines += "\r\n"
            self.pendingConnections[requestId] = pending

            guard let headerData = headerLines.data(using: .utf8) else { return }
            pending.connection.send(content: headerData, completion: .contentProcessed({ _ in }))
        }
    }

    func sendResponseData(requestId: String, chunkBase64: String, index: Int) {
        queue.async { [weak self] in
            guard let self, let pending = self.pendingConnections[requestId] else { return }
            guard var data = Data(base64Encoded: chunkBase64) else { return }

            if pending.useChunked {
                let hexSize = String(format: "%X", data.count)
                var chunked = Data()
                chunked.append(hexSize.data(using: .utf8)!)
                chunked.append(Data([0x0D, 0x0A])) // \r\n
                chunked.append(data)
                chunked.append(Data([0x0D, 0x0A])) // \r\n
                data = chunked
            }

            pending.connection.send(content: data, completion: .contentProcessed({ _ in }))
        }
    }

    func sendResponseEnd(requestId: String) {
        queue.async { [weak self] in
            guard let self, let pending = self.pendingConnections.removeValue(forKey: requestId) else { return }

            if pending.useChunked {
                let finalChunk = "0\r\n\r\n".data(using: .utf8)!
                pending.connection.send(content: finalChunk, completion: .contentProcessed({ _ in
                    pending.connection.send(content: nil, contentContext: .finalMessage, isComplete: true, completion: .contentProcessed({ _ in
                        pending.connection.cancel()
                    }))
                }))
            } else {
                pending.connection.send(content: nil, contentContext: .finalMessage, isComplete: true, completion: .contentProcessed({ _ in
                    pending.connection.cancel()
                }))
            }
        }
    }

    func sendResponseError(requestId: String, errorMessage: String) {
        queue.async { [weak self] in
            guard let self, let pending = self.pendingConnections.removeValue(forKey: requestId) else { return }
            let body = errorMessage.data(using: .utf8) ?? Data()
            let response = "HTTP/1.1 502 Bad Gateway\r\nContent-Type: text/plain\r\nContent-Length: \(body.count)\r\nConnection: close\r\n\r\n"
            guard let headerData = response.data(using: .utf8) else {
                pending.connection.cancel()
                return
            }
            pending.connection.send(content: headerData + body, completion: .contentProcessed({ _ in
                pending.connection.cancel()
            }))
        }
    }

    // MARK: - Connection Handling

    private func handleConnection(_ connection: NWConnection, webView: WKWebView) {
        connection.start(queue: queue)
        readHTTPRequest(connection, webView: webView, accumulatedData: Data())
    }

    private func handleRawTunnelConnection(
        _ connection: NWConnection,
        targetHost: String,
        port: Int,
        webView: WKWebView
    ) {
        let id = "tcp_\(UUID().uuidString.replacingOccurrences(of: "-", with: ""))"
        let tunnel = RawTunnel(
            id: id,
            port: port,
            connection: connection,
            webView: webView
        )
        rawTunnels[id] = tunnel
        connection.stateUpdateHandler = { [weak self, weak tunnel] state in
            guard let self, let tunnel else { return }
            switch state {
            case .failed(let error):
                self.closeRawTunnel(tunnel, notifyJavaScript: true, error: error.localizedDescription)
            case .cancelled:
                self.closeRawTunnel(tunnel, notifyJavaScript: true)
            default:
                break
            }
        }
        connection.start(queue: queue)
        callProxy(
            webView,
            function: "tcpOpen",
            arguments: ["id": id, "host": targetHost, "port": port]
        )
    }

    private func receiveRawTunnelData(_ tunnel: RawTunnel) {
        guard !tunnel.closed, !tunnel.reading, tunnel.sendCredit > 0 else { return }
        tunnel.reading = true
        let maximum = min(Self.tunnelFrameBytes, tunnel.sendCredit)
        tunnel.connection.receive(
            minimumIncompleteLength: 1,
            maximumLength: maximum
        ) { [weak self, weak tunnel] data, _, isComplete, error in
            guard let self, let tunnel else { return }
            tunnel.reading = false
            if let error {
                self.closeRawTunnel(tunnel, notifyJavaScript: true, error: error.localizedDescription)
                return
            }
            if let data, !data.isEmpty {
                tunnel.sendCredit -= data.count
                self.callProxy(
                    tunnel.webView,
                    function: "tcpData",
                    arguments: ["id": tunnel.id, "chunk": data.base64EncodedString()]
                )
            }
            if isComplete {
                self.callProxy(
                    tunnel.webView,
                    function: "tcpEnd",
                    arguments: ["id": tunnel.id]
                )
                return
            }
            self.receiveRawTunnelData(tunnel)
        }
    }

    private func closeRawTunnel(
        _ tunnel: RawTunnel,
        notifyJavaScript: Bool,
        error: String? = nil
    ) {
        guard !tunnel.closed else { return }
        tunnel.closed = true
        rawTunnels.removeValue(forKey: tunnel.id)
        tunnel.connection.cancel()
        if notifyJavaScript {
            callProxy(
                tunnel.webView,
                function: "tcpClose",
                arguments: ["id": tunnel.id]
            )
        }
        if let error, !error.isEmpty {
            print("[EmbeddedProxy] TCP tunnel \(tunnel.id) closed: \(error)")
        }
    }

    private func callProxy(
        _ webView: WKWebView?,
        function: String,
        arguments: [String: Any]
    ) {
        guard let webView else { return }
        DispatchQueue.main.async {
            webView.callAsyncJavaScript(
                "window.__shellularProxy && window.__shellularProxy[function](...args)",
                arguments: [
                    "function": function,
                    "args": Self.proxyArguments(function: function, values: arguments),
                ],
                in: nil,
                in: .page
            ) { _ in }
        }
    }

    private static func proxyArguments(function: String, values: [String: Any]) -> [Any] {
        switch function {
        case "tcpOpen":
            return [values["id"] ?? "", values["host"] ?? "localhost", values["port"] ?? 0]
        case "tcpData":
            return [values["id"] ?? "", values["chunk"] ?? ""]
        case "tcpConsumed":
            return [values["id"] ?? "", values["bytes"] ?? 0]
        default:
            return [values["id"] ?? ""]
        }
    }

    private func readHTTPRequest(_ connection: NWConnection, webView: WKWebView, accumulatedData: Data) {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 65536) { [weak self] data, _, isComplete, error in
            guard let self else { return }

            if error != nil {
                connection.cancel()
                return
            }

            guard let data = data, !data.isEmpty else {
                if isComplete { connection.cancel() }
                else { self.readHTTPRequest(connection, webView: webView, accumulatedData: accumulatedData) }
                return
            }

            var allData = accumulatedData
            allData.append(data)

            let request = self.parseHTTPRequest(allData)
            guard let req = request else {
                if isComplete {
                    self.sendErrorResponse(connection, status: 400, message: "Bad Request")
                } else {
                    self.readHTTPRequest(connection, webView: webView, accumulatedData: allData)
                }
                return
            }

            let contentLength = req.headers.first(where: { $0.key.lowercased() == "content-length" })
                .flatMap { Int($0.value) } ?? 0
            if contentLength > 0, let body = req.body {
                if body.count < contentLength && !isComplete {
                    self.readHTTPRequest(connection, webView: webView, accumulatedData: allData)
                    return
                }
            }

            self.proxyRequest(connection, request: req, webView: webView)
        }
    }

    private func proxyRequest(_ connection: NWConnection, request: ParsedRequest, webView: WKWebView) {
        let requestId = "req_\(UUID().uuidString.replacingOccurrences(of: "-", with: ""))"

        queue.async { [weak self] in
            self?.pendingConnections[requestId] = PendingConnection(connection: connection, requestId: requestId)
        }

        DispatchQueue.main.async {
            let headersDict = Dictionary(uniqueKeysWithValues: request.headers.map { ($0.key, $0.value) })
            let headersJson = (try? JSONSerialization.data(withJSONObject: headersDict)).flatMap {
                String(data: $0, encoding: .utf8)
            } ?? "{}"

            let bodyStr = request.body.flatMap { String(data: $0, encoding: .utf8) } ?? ""

            let escapedHeaders = self.escapeForJS(headersJson)
            let escapedBody = self.escapeForJS(bodyStr)
            let escapedUrl = self.escapeForJS(request.url)
            let escapedMethod = self.escapeForJS(request.method)

            let js = "window.__shellularProxy&&window.__shellularProxy.httpRequest('\(requestId)','\(escapedMethod)','\(escapedUrl)','\(escapedHeaders)','\(escapedBody)')"
            webView.evaluateJavaScript(js)
        }
    }

    private func escapeForJS(_ str: String) -> String {
        return str
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "'", with: "\\'")
            .replacingOccurrences(of: "\n", with: "\\n")
            .replacingOccurrences(of: "\r", with: "\\r")
    }

    // MARK: - Error Response

    private func sendErrorResponse(_ connection: NWConnection, status: Int, message: String) {
        let body = message.data(using: .utf8) ?? Data()
        let response = "HTTP/1.1 \(status) Error\r\nContent-Type: text/plain\r\nContent-Length: \(body.count)\r\nConnection: close\r\n\r\n"
        guard let headerData = response.data(using: .utf8) else {
            connection.cancel()
            return
        }
        connection.send(content: headerData + body, completion: .contentProcessed({ _ in
            connection.cancel()
        }))
    }

    // MARK: - HTTP Parsing

    struct ParsedRequest {
        let method: String
        let url: String
        let headers: [(key: String, value: String)]
        let body: Data?
    }

    private func parseHTTPRequest(_ data: Data) -> ParsedRequest? {
        guard let string = String(data: data, encoding: .utf8) ?? String(data: data, encoding: .ascii) else {
            return nil
        }

        let parts = string.components(separatedBy: "\r\n\r\n")
        guard let headerSection = parts.first else { return nil }

        let bodyStr = parts.count > 1 ? parts.dropFirst().joined(separator: "\r\n\r\n") : nil
        let body = bodyStr.flatMap { $0.data(using: .utf8) }

        var lines = headerSection.components(separatedBy: "\r\n")
        guard let requestLine = lines.first else { return nil }
        lines.removeFirst()

        let requestParts = requestLine.components(separatedBy: " ")
        guard requestParts.count >= 2 else { return nil }

        let method = requestParts[0].uppercased()
        let rawUrl = requestParts[1]

        var headers: [(key: String, value: String)] = []
        for line in lines {
            guard let colonIndex = line.firstIndex(of: ":") else { continue }
            let key = String(line[..<colonIndex]).trimmingCharacters(in: .whitespaces)
            let value = String(line[line.index(after: colonIndex)...]).trimmingCharacters(in: .whitespaces)
            headers.append((key: key, value: value))
        }

        return ParsedRequest(method: method, url: rawUrl, headers: headers, body: body)
    }
}
